src/config/base.cpp

```diff
@@ -656,41 +1030,63 @@ LIBSESSION_EXPORT config_push_data* config_push(config_object* conf) {
+    // We need to do one alloc here that holds everything.  We pack it as follows:
     // - the returned struct
-    // - pointers to the obsolete message hash strings
-    // - the data
-    // - the message hash strings
-    size_t buffer_size = sizeof(config_push_data) + obs.size() * sizeof(char*) + data.size();
+    // - data pointers: [*configdata1][*configdata2]...  <-- `config` points to the beginning of
+    // this
+    // - size_t [size1][size2]... <-- `config_lens` points to the beginning of this
+    // - obsolete hash pointers: [*obs1][*obs2]...  <-- `obsolete` points to the beginning of this
+    // - data: [configdata1][configdata2]...[obs1\0][obs2\0]...
+    static_assert(alignof(config_push_data) >= alignof(char*));
+    static_assert(sizeof(config_push_data) % alignof(char*) == 0);
+    static_assert(alignof(char*) == alignof(size_t*));
+    static_assert(alignof(size_t) == alignof(char*));
+    size_t buffer_size = sizeof(config_push_data) + data.size() * (sizeof(char*) + sizeof(size_t)) +
+                         obs.size() * sizeof(char*);
+    for (auto& d : data)
+        buffer_size += d.size();
     for (auto& o : obs)
         buffer_size += o.size();
     buffer_size += obs.size();  // obs msg hash string NULL terminators
 
     auto* ret = static_cast<config_push_data*>(std::malloc(buffer_size));
+    if (!ret) {
+        // TODO: uncomment this when we start using oxen-logging:
+        // log::critical(logcat, "Memory allocation failed in config_push!");
+        return nullptr;
+    }
 
     ret->seqno = seqno;
-
-    static_assert(alignof(config_push_data) >= alignof(char*));
-    ret->obsolete = reinterpret_cast<char**>(ret + 1);
+    ret->config = reinterpret_cast<unsigned char**>(ret + 1);
+    ret->config_lens = reinterpret_cast<size_t*>(ret->config + 1);
+    ret->n_configs = data.size();
+    ret->obsolete = reinterpret_cast<char**>(ret->config_lens + 1);
     ret->obsolete_len = obs.size();
 
-    ret->config = reinterpret_cast<unsigned char*>(ret->obsolete + ret->obsolete_len);
-    ret->config_len = data.size();
-
-    std::memcpy(ret->config, data.data(), data.size());
-    char* obsptr = reinterpret_cast<char*>(ret->config + ret->config_len);
+    unsigned char* pos = reinterpret_cast<unsigned char*>(ret->obsolete + 1);
+    for (size_t i = 0; i < data.size(); i++) {
+        std::memcpy(pos, data[i].data(), data[i].size());
+        ret->config[i] = pos;
+        pos += (ret->config_lens[i] = data[i].size());
+    }
     for (size_t i = 0; i < obs.size(); i++) {
-        std::memcpy(obsptr, obs[i].c_str(), obs[i].size() + 1);
-        ret->obsolete[i] = obsptr;
-        obsptr += obs[i].size() + 1;
+        auto cstr_len = obs[i].size() + 1 /*NUL terminator*/;
+        std::memcpy(pos, obs[i].c_str(), cstr_len);
+        ret->obsolete[i] = reinterpret_cast<char*>(pos);
+        pos += cstr_len;
     }
+    assert(pos - reinterpret_cast<unsigned char*>(ret) == buffer_size);
```

Comment:
This assertion seems to be getting tripped during the first `config_push(conf)` call iOS makes after creating a new account (for the `UserProfile` config)

The dump for the config is: `64313a21693165313a2439313a64313a23693165313a2664313a2b692d3165313a6e31333a546573744d756c7469436f6e6665313a3c6c6c69306533323aea173b57beca8af18c3519a7bbf69c3e7a05d1c049fa9558341d8ebb48b0c96564656565313a3d646565313a286c65313a296c65313a2a646565`

And the secret key for this empty test account is: `02b272b5eac54c3f67071fdc688fbbbe00000000000000000000000000000000e7efe716cbcf992e57e2c6e94df74b2eddce2074272aca1e5038e5d13b21f715`

Let me know if you need any more info to help debug 🙂

Reply:
Fixed.